src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/SelectExpandNode.cs

```diff
@@ -242,30 +253,125 @@ public SelectExpandNode(SelectExpandClause selectExpandClause, IEdmStructuredTyp
                         }
                         else
                         {
                             ReferencedNavigationProperties.Add(navigationProperty);
                         }
                     }
+                    else
+                    {
+                        //This is the case where the navigation property is not on the current type. We need to propagate the expand item to deeper SelectExpandNode.
+                        IEdmProperty complexProperty = FindNextPropertySegment(expandReferenceItem.PathToNavigationProperty);
+
+                        if (complexProperty != null && ExpandedNavigationPropertiesOnComplexTypes.ContainsKey(complexProperty))
+                        {
+                            SelectExpandClause oldClause = ExpandedNavigationPropertiesOnComplexTypes[complexProperty].SelectAndExpand;
+                            SelectExpandClause newClause = new SelectExpandClause(oldClause.SelectedItems.Concat(new List<SelectItem>() { expandReferenceItem }), false);
+                            ExpandedNavigationSelectItem newItem = new ExpandedNavigationSelectItem(expandReferenceItem.PathToNavigationProperty, navigationSegment.NavigationSource, newClause);
+                            ExpandedNavigationPropertiesOnComplexTypes.Add(complexProperty, newItem);
+                        }
+                        else if (complexProperty != null)
+                        {
+                            List<SelectItem> expandedItemsForClause = new List<SelectItem>();
+                            expandedItemsForClause.Add(expandReferenceItem);
+                            SelectExpandClause clause = new SelectExpandClause(expandedItemsForClause, false);
+                            ExpandedNavigationSelectItem expandItem = new ExpandedNavigationSelectItem(expandReferenceItem.PathToNavigationProperty, navigationSegment.NavigationSource, clause);
+                            ExpandedNavigationPropertiesOnComplexTypes.Add(complexProperty, expandItem);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Finds the appropriate property segment which should be responsible for propagating the expand item.
+        /// For instance, if we are creating the SelectExpandNode for property p2 for the query ~/EnitySet/Key/p1/p2/p3?$expand=NP, we want to return property p3 here. 
+        /// </summary>
+        private IEdmStructuralProperty FindNextPropertySegment(ODataPath path)
+        {
+            IEdmStructuralProperty complexProperty = null;
+            // If the current SelectExpandNode is not top-level and has a property associated with it then return the next property segment from the path.
+            if (Property != null)
+            {
+                int pathIndex = 0;
+                bool endOfPathToProperty = false;
+                IList<string> pathSegments = PathToProperty.PathSegments.ToList();
```

Comment:
>PathToProperty [](start = 45, length = 14)

Any chance that PathToProperty might be null here?  What if a user sets Property but not PathToProperty? Should probably handle to avoid a null ref exception.